src/main/kotlin/com/github/quiltservertools/ledger/utility/NbtUtils.kt

```diff
@@ -21,4 +38,98 @@ object NbtUtils {
 
         return NbtHelper.toBlockState(stateTag)
     }
+
+    fun itemToProperties(item: ItemStack): NbtCompound? {
+        val itemTag = NbtCompound()
+
+        val storedNBT = item.writeNbt(NbtCompound())
+
+        Ledger.logger.info(storedNBT)
+
+        if (item.count > 1) { itemTag.putByte(COUNT, item.count.toByte()) }
+        if (item.nbt != null) { itemTag.put(TAG, item.nbt) }
+
+        return if (itemTag.isEmpty) null else itemTag
+    }
+
+    fun itemFromProperties(tag: String?, name: Identifier): ItemStack {
+        val itemTag = NbtCompound()
+
+        itemTag.putString("id", name.toString())
+
+        if (tag == null) {
+            itemTag.putByte(COUNT, 1);
+            return ItemStack.fromNbt(itemTag)
+        }
+
+        val tagNbt = StringNbtReader.parse(tag)
+
+        if (tagNbt.contains(COUNT)) { itemTag.putByte(COUNT, tagNbt.getByte(COUNT)) }
+        else { itemTag.putByte(COUNT, 1) }
+
+        if (tagNbt.contains(TAG)) { itemTag.put(TAG, tagNbt.getCompound(TAG)) }
+
+        return ItemStack.fromNbt(itemTag)
+    }
+
+    //This monster will find non default NBT tags & log them
+    fun entityToProperties(entity: Entity, actionType: String): NbtCompound? {
```

Comment:
Why is this needed?

Any unused NBT tags (at least in vanilla) will do nothing if there's no functionality for them. There's no need to validate any NBT

Reply:
Originally the plan was to try to reduce how much was being stored to the db, armor stands mainly get quite large when fully equipped with enchanted armor and swapping items would normally mean the entire entity and the item get logged, when for those actions really all that is needed is the UUID and the item.

wanted to see if it could be applied to other parts that use extra data like items and entities that died but eh, it's still pretty rough around the edges for that, the checks are insane, I think I was just overly concerned if somehow a property for an entity got onto an item.